src/protocol/boolean/bit_decomposition.rs

```diff
@@ -61,16 +61,20 @@ impl BitDecomposition {
         .await?;
 
         // Step 7. a bitwise scalar value `f_B = bits(2^l - p)`
-        let l = F::Integer::BITS;
+        let l = u128::BITS - F::PRIME.into().leading_zeros();
         let x = 2_u128.pow(l) - F::PRIME.into();
         let f_b = (0..l).map(|i| F::from(x >> i & 1));
 
         // Step 8, 9. [g_i] = [q] * f_i
-        let g_b = f_b
+        let mut g_b = f_b
             .into_iter()
             .map(|f_bit| q_p.clone() * f_bit)
             .collect::<Vec<_>>();
 
+        // Since bitwise-sum will return +1 bit from the carry, d_b.len() is 1 bit longer
+        // than g_b. Append ZERO to make them the same size.
+        g_b.push(S::ZERO);
```

Comment:
Rather than this, you can use `.chain(std::iter::once())` before collecting.  This will result in the allocation for the vector being correct.